kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -200,10 +201,23 @@ protected CloseableIterator<FilteredColumnarBatch> getScanFiles(
 
   @Override
   public Row getScanState(Engine engine) {
+    // Check for metadata columns in the read schema and convert them to physical columns
+    List<StructField> convertedFields = new ArrayList<>();
+    for (StructField field : readSchema.fields()) {
+      Optional<StructField> converted =
+          MaterializedRowTrackingColumn.convertToPhysicalColumn(field, metadata);
+      if (converted.isPresent()) {
+        convertedFields.add(converted.get());
+      } else {
+        convertedFields.add(field);
+      }
+    }
+    StructType convertedSchema = new StructType(convertedFields);
```

Comment:
Move all of this logic inside `convertToPhysicalColumn`, that then directly returns a `StructType` with row tracking columns replaced.

We can pick between two options, either:
1. We consider this logic to be specific to row tracking metadata columns, and everything should be handled in convertToPhysicalColumn.
2. We consider this logic to be more generic for metadata columns in general, in that case `convertToPhysicalColumn` should be defined here in ScanImpl.java and be written more generically, e.g. by using a `switch` over supported metadata columns and failing if we don't recognize one.

I think 1. is fine for now, we don't really have evidence at this stage that we'll add metadata columns that map to a physical column beyond row tracking